TimeManagementService_WorkTime_Create

KostinP opened the pull request:
It was necessary to add the ability to add working hours. For this, I added:
- Repository
- Request model
- Mapper
- Validator
- Command
- Controller
- Tests (**_something is wrong_**)

KostinP committed:
Added the main part of the task "Create WorkTime"

KostinP commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,80 @@
+                .NotEmpty();
+        }
+
+        private bool TimeAndDateOfNewWorkTimeIsValid(
```
This should not be a separate method, because it will not produce specialized error messages, I will redo.

KostinP replied:
And it would be nice to add comments.

KostinP replied:
Not whole numbers are compared. Apparently you need to use a interval.

KostinP replied:
Fixed.

KostinP commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,80 @@
+    {
+        private IWorkTimeRepository repository;
+
+        // How many days ago can WorkTime be filled.
```
Variables used should not be private, but public and immutable.

KostinP replied:
And there is not enough restriction on working in a row.

KostinP replied:
Not whole numbers are compared. Apparently you need to use a interval.

KostinP replied:
Fixed.

KostinP committed:
Fixed validator

DarkFelas commented on DigitalOffice/Services/TimeManagementService/Commands/CreateWorkTimeCommand.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            validator.ValidateAndThrow(request);
+
+            var workTime = mapper.Map(request);
```
rename dbWorkTime

KostinP replied:
Fixed.

KostinP committed:
Added base of tests

KostinP committed:
Merge branch 'develop' into feature/TimeManagementService_WorkTime_Create

KostinP committed:
Resolved conflicts

KostinP committed:
Fixed name

KostinP committed:
Resolved conflicts (in tests)

denis-kriv commented on DigitalOffice/Services/TimeManagementService/Controllers/TimeManagementController.cs:
```diff
@@ -1,11 +1,24 @@
     public class TimeManagementController : ControllerBase
     {
-        //TODD Add endpoints.
+        public TimeManagementController()
+        {
+        }
```
do we have this?

KostinP replied:
Fixed.

denis-kriv commented on DigitalOffice/Tests/TimeManagementServiceTests/Commands/CreateWorkTimeCommandTests.cs:
```diff
@@ -0,0 +1,43 @@
+            command = new CreateWorkTimeCommand(validatorMock.Object, mapperMock.Object, repositoryMock.Object);
+        }
+
+        [Test]
+        public void ATest()
+        {
+            Assert.Pass();
+        }
+
+        [Test]
+        public void BTest()
+        {
+            Assert.Pass();
+        }
```
I can not understand what it is.

KostinP replied:
In a comment to the pull-request, I wrote that I have not done it yet.

KostinP replied:
Fixed.

denis-kriv commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,31 @@
+            mapper = new WorkTimeMapper();
+        }
+
+        [Test]
+        public void ATest()
+        {
+            Assert.Pass();
+        }
+
+        [Test]
+        public void BTest()
+        {
+            Assert.Pass();
+        }
```
I can not understand

KostinP replied:
Fixed.

denis-kriv commented on DigitalOffice/Tests/TimeManagementServiceTests/Repositories/WorkTimeRepositoryTests.cs:
```diff
@@ -0,0 +1,57 @@
+
+        //CreateWorkTimeTests
+
+        [Test]
+        public void ATest()
+        {
+            Assert.Pass();
+        }
+
+        [Test]
+        public void BTest()
+        {
+            Assert.Pass();
+        }
+
+        //GetUserWorkTimesTests
+
+        [Test]
+        public void CTest()
+        {
+            Assert.Pass();
+        }
```
And this

KostinP replied:
Fixed.

denis-kriv commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,36 @@
+            validator = new CreateWorkTimeRequestValidator(repositoryMock.Object);
+        }
+
+        [Test]
+        public void ATest()
+        {
+            Assert.Pass();
+        }
+
+        [Test]
+        public void BTest()
+        {
+            Assert.Pass();
+        }
```
and this

KostinP replied:
Fixed, but tests not work.

KostinP committed:
Fixed solution

- Changed Comment to Description
Adding comments will be the responsibility of another service
- Removed empty constructor
- Added rule for Title
- Added mapper tests

KostinP committed:
Changed bool to Guid

KostinP committed:
Updated mapper tests

KostinP committed:
Added command tests

djino97 commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,92 @@
+                    {
+                        var startDate = oldWT.Date;
+                        var endDate = oldWT.Date.AddMinutes(oldWT.Time.TotalMinutes);
+                        if (endTimeOfNewWT <= startDate) { } // First new WorkTime, then old workTime
```
Add one line

KostinP replied:
Fixed.

djino97 commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,92 @@
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                })
```
remove line

KostinP replied:
line 80?

djino97 commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,76 @@
+
+            Assert.IsNotNull(result);
+            Assert.IsInstanceOf<DbWorkTime>(result);
+            Assert.AreEqual(FakeProjectId, result.ProjectId);
+            Assert.AreEqual(FakeDate, result.Date);
+            Assert.AreEqual(FakeTime, result.Time);
+            Assert.AreEqual(FakeTitle, result.Title);
+            Assert.AreEqual(FakeDescription, result.Description);
+            Assert.AreEqual(FakeWorkerUserId, result.WorkerUserId);
```
Add SerializerAssert, serialize objects to the json objects inside methods of class SerializerAsser and further compare in one line.

KostinP replied:
Fixed.

djino97 commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,76 @@
+
+            Assert.IsNotNull(result);
+            Assert.IsInstanceOf<DbWorkTime>(result);
+            Assert.AreEqual(FakeProjectId, result.ProjectId);
+            Assert.AreEqual(FakeDate, result.Date);
+            Assert.AreEqual(FakeTime, result.Time);
+            Assert.AreEqual(FakeTitle, result.Title);
+            Assert.IsTrue(string.IsNullOrEmpty(result.Description));
+            Assert.AreEqual(FakeWorkerUserId, result.WorkerUserId);
```
add SerializerAssert

KostinP replied:
Fixed.

KostinP committed:
Corrections after review @djino97

MihailKulikov reviewed (changes requested):
Waiting for unit tests

KostinP committed:
Added other tests (something is wrong)

lalagvana reviewed (changes requested):
Merge from develop, please. There are already other namespaces.

KostinP commented:
> Merge from develop, please. There are already other namespaces.

Ok. But I need more help with tests. Problem with tests: the validator reaches to add the clock to the date, and if the date is not specified, then it crashes with an error. So the question is, can the validator fail with this error, or should we add a check for null?

I will add the appropriate label.

KostinP committed:
Merge branch 'develop' into feature/TimeManagementService_WorkTime_Create

KostinP committed:
Resolve conflicts after merge

KostinP committed:
Delete DigitalOffice.sln

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Commands/CreateWorkTimeCommand.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        private IValidator<CreateWorkTimeRequest> validator;
+        private IMapper<CreateWorkTimeRequest, DbWorkTime> mapper;
+        private IWorkTimeRepository repository;
```
Make fields readonly

KostinP replied:
Fixed.

djino97 commented on DigitalOffice/Tests/TimeManagementServiceTests/Repositories/WorkTimeRepositoryTests.cs:
```diff
@@ -0,0 +1,123 @@
+        }
+
+        #region CreateWorkTimeTests
+        [Test]
+        public void SuccessfulAddNewWorkTimeInDb()
+        {
```
At the end of the test name should be Test (convention), do it to the all test.

KostinP replied:
Resolved in Slack.

djino97 commented on DigitalOffice/Tests/TimeManagementServiceTests/Repositories/WorkTimeRepositoryTests.cs:
```diff
@@ -0,0 +1,123 @@
+            Title = $"WorkTime",
+            WorkerUserId = worker1,
+            ProjectId = project2,
+            Date = DateTime.Now.AddDays(-0.7), // 0.7 days ago
+            Time = new TimeSpan(6, 0, 0) // 0.45 days ago
+        };
+
+        [SetUp]
+        public void Setup()
+        {
+            var dbOptions = new DbContextOptionsBuilder<TimeManagementDbContext>()
+                                    .UseInMemoryDatabase(databaseName: "InMemoryDatabase")
+                                    .Options;
+            dbContext = new TimeManagementDbContext(dbOptions);
+            repository = new WorkTimeRepository(dbContext);
+        }
+
+        [TearDown]
+        public void Clear()
+        {
```
What is cleared ?

KostinP replied:
DB. No?

mekk1t replied:
Agree on the naming. Should clarify what exactly do we clear (yeah I see that it uses DbContext so its a DB cleanup)
But naming conventions and overall method readability)

djino97 replied:
Write CleanDb

KostinP replied:
Renamed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/TimeManagementServiceUnitTests.csproj:
```diff
@@ -4,16 +4,18 @@
     <TargetFramework>netcoreapp3.1</TargetFramework>
 
     <IsPackable>false</IsPackable>
-
-    <AssemblyName>LT.DigitalOffice.TimeManagementServiceUnitTests</AssemblyName>
-
-    <RootNamespace>LT.DigitalOffice.TimeManagementServiceUnitTests</RootNamespace>
```
It must be returned, otherwise the namespace will be written incorrectly

KostinP replied:
When I put this code back, the solution stops loading. I'll try again tomorrow.

MihailKulikov replied:
I had no problem. Something like this should turn out
![изображение](https://user-images.githubusercontent.com/61084995/89121313-7fbe1c00-d4c6-11ea-83f7-2284802e3720.png)

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,251 @@
+                Description = Description
+            };
+
+            var sucessfulResult = validator.Validate(successfulRequest);
```
Typo

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Mappers/WorkTimeMapper.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public class WorkTimeMapper : IMapper<CreateWorkTimeRequest, DbWorkTime>
+    {
+        public DbWorkTime Map(CreateWorkTimeRequest request)
```
Add checking for null

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Repositories/WorkTimeRepository.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    public class WorkTimeRepository : IWorkTimeRepository
+    {
+        private TimeManagementDbContext dbContext;
```
Add readonly

KostinP replied:
Fixed.

djino97 commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,81 @@
+        [Test]
+        public void ReturnRightModelSuccesfully()
+        {
+            var expectedWorkTime = new DbWorkTime
+            {
+                ProjectId = request.ProjectId,
+                Date = request.Date,
+                Time = request.Time,
+                Title = request.Title,
+                Description = request.Description,
+                WorkerUserId = request.WorkerUserId
+            };
```
Move this object to the SetUp method

KostinP replied:
Fixed.

djino97 commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,81 @@
+        public void ReturnRightModelWithNullDescriptionSuccesfully()
+        {
+            request.Description = null;
+            var expectedWorkTime = new DbWorkTime
+            {
+                ProjectId = request.ProjectId,
+                Date = request.Date,
+                Time = request.Time,
+                Title = request.Title,
+                WorkerUserId = request.WorkerUserId
+            };
```
Move this object to the SetUp method

KostinP replied:
Fixed.

djino97 commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,81 @@
+        private IMapper<CreateWorkTimeRequest, DbWorkTime> mapper;
+
+        private CreateWorkTimeRequest request;
+
+        [SetUp]
+        public void SetUp()
```
[OneTimeSetup]

KostinP replied:
I use the fact that it fills up every time.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Repositories/WorkTimeRepository.cs:
```diff
@@ -0,0 +1,54 @@
+                return dbContext.WorkTimes.Where(predicate).ToList();
+            }
+
+            if (filter.EndTime != null)
```
DateTime can not be null.

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Repositories/WorkTimeRepository.cs:
```diff
@@ -0,0 +1,54 @@
+                predicate.And(wt => wt.Date >= filter.StartTime);
+            }
+
+            if (filter.StartTime != null)
```
DateTime can not be null

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Commands/CreateWorkTimeCommandTests.cs:
```diff
@@ -0,0 +1,101 @@
+                .Setup(x => x.Validate(It.IsAny<IValidationContext>()).IsValid)
+                .Returns(false);
+
+            Assert.Throws<ValidationException>(() => command.Execute(request));
```
Also it will be cool to verify that method of repositoryMock is not called

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Commands/CreateWorkTimeCommandTests.cs:
```diff
@@ -0,0 +1,101 @@
+        }
+
+        [Test]
+        public void SuccesfulCreateNewWorkTimer()
```
Typo

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Commands/CreateWorkTimeCommandTests.cs:
```diff
@@ -0,0 +1,101 @@
+                .Setup(x => x.CreateWorkTime(It.IsAny<DbWorkTime>()))
+                .Returns(createdWorkTime.Id);
+
+            Assert.AreEqual(createdWorkTime.Id, command.Execute(request));
```
Also it will be cool to verify that method of repositoryMock is called

KostinP replied:
Fixed.

mekk1t replied:
Why not to verify all the subsequent mocks?

mekk1t replied:
Like that: 
![image](https://user-images.githubusercontent.com/61093487/89462967-dea9bc80-d776-11ea-877e-e004c10f2b0c.png)

mekk1t replied:
One note: I'm not sure I'm verifying validators right.
It throws me an exception when I write <...> Validate(request), so instead I wrote in the verification that validator validates any IValidationContext supported objects types

mekk1t replied:
Ah, I see. You verified repo like i did the validator. Alright, will use this in future tests as well (instead of passing actual arguments).
Or actual arguments are also good?

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        [Test]
+        public void ReturnRightModelSuccesfully()
```
typo

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,81 @@
+            };
+
+            var newWortTime = mapper.Map(request);
+            expectedWorkTime.Id = newWortTime.Id;
```
Move to SetUp

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,81 @@
+                Time = request.Time,
+                Title = request.Title,
+                WorkerUserId = request.WorkerUserId
+            };
```
Move to SetUp

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Repositories/WorkTimeRepositoryTests.cs:
```diff
@@ -0,0 +1,123 @@
+        [Test]
+        public void SuccessfulAddNewWorkTimeInDb()
+        {
+            var guidOfNewWT = repository.CreateWorkTime(testWorkTime1);
```
This is not so important, but it is better to name the variables without abbreviations. Using modern IDE, there should be no problem with long variable names

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+        /// <summary>
+        /// How many days ago can WorkTime be added.
+        /// </summary>
+        public readonly static int fromDay = 3;
+        /// <summary>
+        /// How many days ahead can WorkTime be added.
+        /// </summary>
+        public readonly static int toDay = 2;
```
Why not a constant?

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+            RuleFor(wt => wt.Date)
+                .NotEmpty()
+                .When(wt => wt.Date > fromDateTime)
+                .WithMessage(wt => string.Format(
```
Use string interpolation expression

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+                    "WorkTime had to be filled no later than {0}.",
+                    fromDateTime.ToString(System.Globalization.CultureInfo.CreateSpecificCulture("en-GB"))))
+                .When(wt => wt.Date < toDateTime)
+                .WithMessage(wt => string.Format(
```
Use string interpolation expression

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+                .When(wt => wt.Time.TotalMinutes <= 0)
+                .WithMessage("You cannot indicate that you worked zero hours or a negative amount.")
+                .When(wt => Math.Round(wt.Time.TotalMinutes, 12) > Math.Round(workingLimit.TotalMinutes, 12))
+                .WithMessage(wt => string.Format(
```
Use string interpolation expression

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+                            EndTime = endTimeOfNewWT.AddMinutes(workingLimit.TotalMinutes)
+                        });
+
+                    foreach (var oldWT in workTimes)
+                    {
+                        var startDate = oldWT.Date;
+                        var endDate = oldWT.Date.AddMinutes(oldWT.Time.TotalMinutes);
+
+                        if (endTimeOfNewWT <= startDate) { } // First new WorkTime, then old workTime
+                        else if (endDate <= startTimeOfNewWT) { } // First old WorkTime, then new WorkTime
+                        else
+                        {
+                            return false;
+                        }
+                    }
```
Can be convert to LINQ expression: 
return (from oldWT in workTimes
                        let startDate = oldWT.Date
                        let endDate = oldWT.Date.AddMinutes(oldWT.Time.TotalMinutes)
                        where endTimeOfNewWT > startDate
                        select endDate).All(endDate => endDate <= startTimeOfNewWT);

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,251 @@
+        {
+            var request = new CreateWorkTimeRequest
+            {
+                //WorkerUserId = WorkerUserId,
```
Very interesting decision. Pls, edit. WorkerUserId = Guid.Empty will be better

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,251 @@
+            var request = new CreateWorkTimeRequest
+            {
+                WorkerUserId = WorkerUserId,
+                //Date = Date,
```
Edit

KostinP replied:
How to do it differently?

MihailKulikov replied:
Date = new DateTime() will be better, I think.

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,251 @@
+            {
+                WorkerUserId = WorkerUserId,
+                Date = Date,
+                //Time = Time,
```
Edit

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,251 @@
+                WorkerUserId = WorkerUserId,
+                Date = Date,
+                Time = Time,
+                //Title = Title,
```
Edit

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+            this.repository = repository;
+
+            RuleFor(wt => wt.WorkerUserId)
+                    .NotEmpty();
```
why is there no message?

KostinP replied:
Because the standard error message contains all the information you need.

MihailKulikov commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,251 @@
+                Date = Date,
+                Time = Time,
+                Title = Title,
+                //ProjectId = ProjectId,
```
Edit

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+                .WithMessage(wt => string.Format(
+                    "You cannot indicate that you worked more than {0} hours and {1} minutes.",
+                    workingLimit.Hours, workingLimit.Minutes))
+                .When(wt =>
+                {
+                    var startTimeOfNewWT = wt.Date;
+                    var endTimeOfNewWT = startTimeOfNewWT.AddMinutes(wt.Time.TotalMinutes);
+
+                    var workTimes = repository.GetUserWorkTimes(
+                        wt.WorkerUserId,
+                        new WorkTimeFilter
+                        {
+                            StartTime = startTimeOfNewWT.AddMinutes(-workingLimit.TotalMinutes),
+                            EndTime = endTimeOfNewWT.AddMinutes(workingLimit.TotalMinutes)
+                        });
+
+                    foreach (var oldWT in workTimes)
+                    {
+                        var startDate = oldWT.Date;
+                        var endDate = oldWT.Date.AddMinutes(oldWT.Time.TotalMinutes);
+
+                        if (endTimeOfNewWT <= startDate) { } // First new WorkTime, then old workTime
+                        else if (endDate <= startTimeOfNewWT) { } // First old WorkTime, then new WorkTime
+                        else
+                        {
+                            return false;
+                        }
+                    }
```
It might be a good idea to move this check into a separate static method. You can see an example in the documentation.
https://docs.fluentvalidation.net/en/latest/custom-validators.html
We need to refactor the code. So far, it's very hard to figure it out.

MihailKulikov replied:
Also, you use When, but should use Must. Check documentation, please.

KostinP replied:
> Also, you use When, but should use Must. Check documentation, please.

Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,93 @@
+
+            RuleFor(wt => wt.Time)
+                .NotEmpty()
+                .When(wt => wt.Time.TotalMinutes <= 0)
```
I think, here TotalMinutes should be > 0.

KostinP replied:
Fixed.

KostinP committed:
Fixed some notes and bugs

KostinP committed:
Fix

KostinP committed:
Update tests

KostinP committed:
Updated .csproj

KostinP committed:
Updated tests

KostinP committed:
Changed Date and Time to StartTime and EndTime

KostinP committed:
Updated tests

Commented

mekk1t commented on DigitalOffice/Services/TimeManagementService/Controllers/TimeManagementController.cs:
```diff
@@ -1,11 +1,21 @@
     public class TimeManagementController : ControllerBase
     {
-        //TODD Add endpoints.
+        [Route("AddWorkTime")]
+        [HttpPost]
+        public Guid AddWorkTime(
```
It's not a suggestion nor a comment, just a thought:
We can also write [HttpPost("AddWorkTime")], that will reduce the code for 1 string.
Spartak approved, so feel free to think about it.

KostinP replied:
Fixed.

mekk1t commented on DigitalOffice/Services/TimeManagementService/Database/Entities/DbLeaveTypes.cs:
```diff
@@ -1,8 +1,7 @@
 {
-   public enum LeaveType
-   {
+    public enum LeaveType
+    {
         Vacation = 0,
         SickLeave = 1,
         Training = 2,
```
The name of the file should be without "S": DbLeaveType.cs, isn't it?

KostinP replied:
Yes. This is needed in another pull requestn other pull request.

mekk1t commented on DigitalOffice/Services/TimeManagementService/Database/TimeManagementDbContext.cs:
```diff
@@ -18,6 +18,5 @@ public TimeManagementDbContext(DbContextOptions<TimeManagementDbContext> options
         protected override void OnModelCreating(ModelBuilder modelBuilder)
         {
         }
```
I think, no matter do we use Fluent API or not, this line should be here:

modelBuilder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());

Anyway, you can argue with me on that. What do you think, should we add it if we don't use it? Common sense tells me not to add it :/

KostinP replied:
Fixed.

mekk1t commented on DigitalOffice/Services/TimeManagementService/Mappers/Interfaces/IMapper.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace LT.DigitalOffice.TimeManagementService.Mappers.Interfaces
+{
+    public interface IMapper<Tin, Tout>
+    {
+        Tout Map(Tin value);
+    }
```
All generics usually (for example, Task that returns "TResult" in Microsoft documentation) are written like TSomething, 
where "Something" starts with an Upper letter. I think we should write everywhere TIn, TOut etc.

KostinP replied:
Fixed.

mekk1t commented on DigitalOffice/Services/TimeManagementService/Repositories/Filters/WorkTimeFilter.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace LT.DigitalOffice.TimeManagementService.Repositories.Filters
+{
+    public class WorkTimeFilter
+    {
+        public DateTime? StartTime { get; set; }
+        public DateTime? EndTime { get; set; }
+    }
```
It is added for the future cases, I suppose?

KostinP replied:
Yes.

mekk1t commented on DigitalOffice/Services/TimeManagementService/Repositories/WorkTimeRepository.cs:
```diff
@@ -0,0 +1,54 @@
+
+        public ICollection<DbWorkTime> GetUserWorkTimes(Guid userId, WorkTimeFilter filter)
+        {
+            var predicate = PredicateBuilder.New<DbWorkTime>();
+
+            predicate.Start(wt => wt.WorkerUserId == userId);
+
+            if (filter == null)
+            {
+                return dbContext.WorkTimes.Where(predicate).ToList();
+            }
+
+            if (filter.StartTime != null)
+            {
+                predicate.And(wt => wt.StartTime >= filter.StartTime);
+            }
+
+            if (filter.EndTime != null)
+            {
+                predicate.And(wt => wt.EndTime <= filter.EndTime);
+            }
```
If filters are gonna be used elsewhere, I think predicate building should be written in another method.

KostinP replied:
This method doesn't seem useful to me. Will take up a lot of space, have a complex signature.

mekk1t replied:
Alright then

mekk1t commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,87 @@
+                    .NotEmpty();
+
+            RuleFor(wt => wt.StartTime)
+                .NotEqual(new DateTime())
+                .Must(st => st > fromDateTime)
+                .WithMessage(date => $"WorkTime had to be filled no later than {fromDateTime.ToString(culture)}.")
+                .Must(st => st < toDateTime)
```
Typo: looks like you mean "wt" as in "WorkTime", not "st";
+ why .NotEqual (new DateTime()) ? Could you please explain? Thank you.

KostinP replied:
No. RuleFor(wt => wt.StartTime), so st.

KostinP replied:
For test Not Empty (work in TMService_LeaveTime_Create)
To prevent the validator from throwing an error once again

mekk1t commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,87 @@
+                .Must(st => st > fromDateTime)
+                .WithMessage(date => $"WorkTime had to be filled no later than {fromDateTime.ToString(culture)}.")
+                .Must(st => st < toDateTime)
+                .WithMessage(date => $"WorkTime cannot be filled until {toDateTime.ToString(culture)}.");
```
Also I'm confused about these messages.
I think you mixed up the first and the second messages.
Like, you can't fill later than ToDateTime (the latest already)
and you can't fill earlier that FromDateTime.

In current state, messages show:
You have to fill time before FromDateTime
You Have to fill time after ToDateTime.

Please, check this. I might be mistaken.

KostinP replied:
No. 
no later than => more then start
until => less then end

mekk1t commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,87 @@
+            RuleFor(wt => wt)
+                .Must(wt => wt.StartTime < wt.EndTime)
+                .WithMessage("You cannot indicate that you worked zero hours or a negative amount.")
+                .Must(wt => wt.StartTime - wt.EndTime <= workingLimit)
+                .WithMessage(time => string.Format(
+                    "You cannot indicate that you worked more than {0} hours and {1} minutes.",
+                    workingLimit.Hours, workingLimit.Minutes))
```
I think this statement is wrong. 
I tested it in a custom console app, and StartTime - EndTime is always negative and so is always less than workingLimit. For that, we should write "EndTime - StartTime"

And, I can't figure this out yet, we should check that the time can be added only from 8 AM to 6PM, for example. Or we can set time like ("I worked from 00:00 to 04:00")?

![image](https://user-images.githubusercontent.com/61093487/89461379-87a2e800-d774-11ea-8aeb-ae3f3225aadb.png)

KostinP replied:
Yes, there is a mistake, it should be the other way around

mekk1t replied:
Above, screenshot of your diff check.
Below: screenshot of should-be diff check. I think this illustrates my point

![image](https://user-images.githubusercontent.com/61093487/89461602-dcdef980-d774-11ea-9dcc-4308d984917f.png)

KostinP replied:
Fixed. Added tests.

mekk1t commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,87 @@
+                        wt.WorkerUserId,
+                        new WorkTimeFilter
+                        {
+                            StartTime = wt.StartTime.AddMinutes(-workingLimit.TotalMinutes),
+                            EndTime = wt.EndTime.AddMinutes(workingLimit.TotalMinutes)
```
Can't figure out why we search for the times 24 hours before and after the one we want to set. 
If we can add times 3 days before and 2 days after (or vice versa), then we should add to StartTime x3 and EndTime x2 modifier? What do you think, Pavel?

KostinP replied:
StartTime. Because this is a limit for any given time.
EndTime. It needs to be removed.

KostinP replied:
Fixed.

mekk1t commented on DigitalOffice/Tests/TimeManagementServiceTests/Commands/CreateWorkTimeCommandTests.cs:
```diff
@@ -0,0 +1,107 @@
+                ProjectId = Guid.NewGuid(),
+                StartTime = new DateTime(2020, 7, 29, 9, 0, 0),
+                EndTime = new DateTime(2020, 7, 29, 17, 0, 0),
+                Title = "I was working on a very important task",
+                Description = "I was asleep. I love sleep. I hope I get paid for this.",
```
💯 xD

mekk1t commented on DigitalOffice/Tests/TimeManagementServiceTests/Commands/CreateWorkTimeCommandTests.cs:
```diff
@@ -0,0 +1,107 @@
+            command = new CreateWorkTimeCommand(validatorMock.Object, mapperMock.Object, repositoryMock.Object);
+        }
+
+        [Test]
+        public void FailCreateNewWorkTimeAccordingToValidator()
+        {
```
So...Do we write "Test" at the end of the name or not? :)

KostinP replied:
Fixed. Not "...Test".

mekk1t commented on DigitalOffice/Tests/TimeManagementServiceTests/Commands/CreateWorkTimeCommandTests.cs:
```diff
@@ -0,0 +1,107 @@
+            Assert.Throws<Exception>(() => command.Execute(request));
+        }
+
+        [Test]
+        public void SuccessfulCreateNewWorkTimer()
+        {
```
Typo: WorkTimeRRR

KostinP replied:
Fixed.

KostinP committed:
Fix

mekk1t commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,72 @@
+        [Test]
+        public void ReturnRightModelSuccessfully()
+        {
+            var newWortTime = mapper.Map(request);
+            expectedDbWorkTimeWithoutId.Id = newWortTime.Id;
```
Typo, worTTTTTT

KostinP replied:
Fixed.

mekk1t commented on DigitalOffice/Tests/TimeManagementServiceTests/Mappers/WorkTimeMapperTests.cs:
```diff
@@ -0,0 +1,72 @@
+            SerializerAssert.AreEqual(expectedDbWorkTimeWithoutId, newWortTime);
+        }
+
+        [Test]
+        public void ReturnRightModelWithNullDescriptionSuccessfully()
+        {
```
Is this a CheckRights or TimeManagement service? haha xD

gotcha, Pavel.
TBH I also messed this up when copypasting some content from ProjectService

KostinP committed:
Fixed typo

mekk1t commented on DigitalOffice/Tests/TimeManagementServiceTests/Validators/CreateWorkTimeRequestValidatorTests.cs:
```diff
@@ -0,0 +1,188 @@
+
+        [Test]
+        public void SuccessfulWorkTimeValidation()
+        {
+            var result = validator.Validate(request);
+            Assert.IsTrue(result.IsValid);
```
What do you think about this writing? 
![image](https://user-images.githubusercontent.com/61093487/89463840-22e98c80-d778-11ea-91a4-73ec4013c866.png)

KostinP replied:
It doesn't matter to me what to use, I'll change it to your version.

KostinP replied:
Fixed.

KostinP committed:
Updated tests

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,89 @@
+        /// <summary>
+        /// How many days ago can WorkTime be added.
+        /// </summary>
+        public const int fromDay = 3;
+        /// <summary>
+        /// How many days ahead can WorkTime be added.
+        /// </summary>
+        public const int toDay = 2;
```
The recommended naming and capitalization convention is to use PascalCasing for constants.

mekk1t replied:
actually, UPPER_SNAKE_CASE;
I think @KostinP  was one of those to suggest that :D

MihailKulikov replied:
UPPER_SNAKE_CASE looks bad

KostinP replied:
When we talked about this, you said that Microsoft uses the proposed top option, can you write examples?

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,89 @@
+        /// <summary>
+        /// Limit on working hours in a row.
+        /// </summary>
+        public readonly static TimeSpan workingLimit = new TimeSpan(24, 0, 0); // 24 hours.
```
Why not get only property?

MihailKulikov replied:
Unhelpful comment

KostinP replied:
> Why not get only property?

?

MihailKulikov replied:
public fields are bad practice. Will be better: 
`public static TimeSpan workingLimit { get; } = new TimeSpan(24, 0, 0);`

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,89 @@
+            RuleFor(wt => wt)
+                .Must(wt => wt.StartTime < wt.EndTime)
+                .WithMessage("You cannot indicate that you worked zero hours or a negative amount.");
+                RuleFor(wt => wt)
```
Are you sure this is needed here?

KostinP replied:
Forgot to remove when I tested

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,89 @@
+                .WithMessage(time => string.Format(
+                    "You cannot indicate that you worked more than {0} hours and {1} minutes.",
+                    workingLimit.Hours, workingLimit.Minutes));
+                RuleFor(wt => wt)
```
Are you sure this is needed here?

KostinP replied:
Forgot to remove when I tested

KostinP replied:
Fixed.

MihailKulikov commented on DigitalOffice/Services/TimeManagementService/Validators/CreateWorkTimeRequestValidator.cs:
```diff
@@ -0,0 +1,89 @@
+
+                    foreach (var oldWorkTime in oldWorkTimes)
+                    {
+                        if (wt.EndTime <= oldWorkTime.StartTime) { }
+                        else if (oldWorkTime.EndTime <= wt.StartTime) { }
+                        else
+                        {
+                            return false;
+                        }
```
Please, refactor this code

KostinP replied:
Fixed.

KostinP committed:
Fix

KostinP committed:
Updated tests

KostinP committed:
Fix

KostinP committed:
Rename Clear to CleanDB

KostinP committed:
Added comments
